File: Tree/102.Binary Tree Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    const stack = [root];
    let count = 1;
    let current = 0;
    let depth = 0;
    const result = [];
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            if (current === 0) {
                result.push([]);
            } 
            result[depth].push(node.val);
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right)
            }
            current++;
            if (current === count) {
                count = stack.length;
                current = 0;
                depth++;
            }
            
        }
    }
    return result;
};

File: Golang/Tree/102.binary-tree-level-order-traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	ret := [][]int{}
	if root != nil {
		stack := []*TreeNode{root}
		length := 1
		for len(stack) > 0 {
			children := []*TreeNode{}
			level := []int{}
			for i := 0; i < length; i++ {
				node := stack[i]
				level = append(level, node.Val)
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			length = len(children)
			stack = children
			ret = append(ret, level)
		}
	}
	return ret
}
